### PR TITLE
SailfishOS support based on SDL2

### PR DESCRIPTION
**Pull request by savegame**
PThis PR add to Godot Engine support for exporting Godot GLES2 games to SailfishOS binary format.  
With this export templates already exists few demos and games:
- [Whoosh! in the Dungeon](https://openrepos.net/content/sashikknox/whoosh-dungeon) - Summary it's just another dungeon crawler!
- [Godot Joypad Demo](https://openrepos.net/content/sashikknox/godot-joypad) for testing gamepads support on SailfishOS (little bit modified for touchscreens)
- [Godot Materials Demo](https://openrepos.net/content/sashikknox/godot-materials-tester) - little bit modified for touchscreens

#### PR bugs fixed :
- ~~clang-fomat code styling~~ **fixed**
- ~~fix glitches on screen rotation~~ **fixed**

~~This pull request fix 26018 issue~~
I am using sources from two pull request 
 - https://github.com/godotengine/godot/pull/16470
 - https://github.com/godotengine/godot/pull/9435  

and build Godot engine for SaiflsihOS. 

#### **What is done**:
- in progress export from editor to RPM (now works only with 3.0.7 SDK release with Saiflsih-3.2.1.20-armv7hl target - becuse its hardcoded ;-) )
- touch screen input
- dynamic orientation events ( Sailfish 3.* have update SDL2 to 2.0.9 :+1: ) 
- dynamic orientation change ( with SDL_SetHint(SDL_HINT_QTWAYLAND_CONTENT_ORIENTATION,"") - not documented in SDL2 wiki, only in **SDL_Hints.h** header )
- normal application life cycle 
- sound initialization
- joypad (just compile already exists **libudev** implementation of LinuxJoypad, works nice )
 
#### **What is not done** (in next update):
- device vibrations 
- use d-bus for disable screen blank (sleep) while engine run in foreground
- ~~pause engiene when application not in foreground~~ **fixed** 
- gyroscope events (not implemented in SDL2 (not try in new SDL2 :question:  ) )
- ~~build RPM with exported project~~ (now i use [make_rpm.sh](https://github.com/savegame/sailfish_godot_rpm/blob/master/make_rpm.sh) script, and [that instructions](https://github.com/savegame/godot/releases/tag/3.2.beta-sailfish-1)  )~~  - **in progress**
- show native keyboard (maliit keyboard) when use text edit UI items 

#### **Known bugs**:
- dont work on Jolla 1, cant debug it, dont have that device 
- ~~have shader compilation error on Jolla Tablet (i try to fix it, now wait test result from users - becuse dont have that device by my self)~~ **fixed**
- SailfishOS platfrom dont support GLES3 for now, waiting for [pull request to libhybris](https://github.com/libhybris/libhybris/pull/361)

#### **Build Instructions**:

For build this branch on SailfishOS, need SailfishSDK (or Sailfish Docker image), all build commands are looks like:
```bash
export SAILFISH_SDK=$HOME/SailfishOS/
export SSH_MER_KEY=${SAILFISH_SDK}/vmshare/ssh/private_keys/engine/mersdk
export MERSSH_PORT=2222
export MERSSH="ssh  -o "IdentitiesOnly=yes" -i \"$SSH_MER_KEY\" -p $MERSSH_PORT"
# GODOT_PATH - path to godot sources inside MerSDK (SailfishSDK) build machine
export GODOT_PATH=/home/mersdk/share/Projects/Sailfish/godot
# version of taget Sailfish platform, should be more than or equal to 3.1.0.12 (becuse of SDL2 version)
export MER_TARGET_VERSION=3.2.0.12
# architecture can be i486 or armv7hl
export MER_TARGET_ARCH=armv7hl
export GODOT_ARCH=`[ "$MER_TARGET_ARCH" == "armv7hl" ] && echo arm || echo x86`
export MER_TARGET=SailfishOS-${MER_TARGET_VERSION}-${MER_TARGET_ARCH}
#first install dependencies
$MERSSH "sb2 -t ${MER_TARGET} -m sdk-install -R zypper in -y SDL2-debugsource SDL2-devel libaudioresource-devel pulseaudio-devel openssl-devel wayland-devel libpng-devel libxkbcommon-devel libudev-devel scons  gcc-c++"
#than build 
$MERSSH "cd $GODOT_PATH && sb2 -t ${MER_TARGET}  scons -j 8 platform=sailfish target=release arch=${GODOT_ARCH} bits=32"
# then optimize binary size if it release (arm or x86)
$MERSSH "cd $GODOT_PATH && sb2 -t ${MER_TARGET}  strip godot.sailfish.opt.${GODOT_ARCH}"
```
and [here](https://github.com/CODeRUS/docker-sailfishos-buildengine) is instructions how use docker with sailfish SDK (thanks to @CODeRUS :+1: )

[Latest build release](https://github.com/savegame/godot/releases/tag/3.1.1-sailfsih-2)

### CODE REVIEW

**Comment by savegame** [comment]
what i need to configure travis build for SailfishOS ?  is it possible install  SailfishSDK on CI server?

**Comment by akien-mga** [comment]
A few quick comments, I haven't reviewed much yet:

- Any PR adding new features should be done against the `master` branch. We're about to release `3.2`, so having this merged in the `3.1` branch would not make sense.
- Bundling the whole of SDL, and that only for a relatively niche platform, is a no-go. I'm sure the SailfishOS SDK provides SDL already to link against. And I assume that you might even be able to dynamically link SDL on SailfishOS so that you don't even need to distribute it, it will use the OS version.

**Comment by savegame** [comment]
Yes, the SDL link dynamically, i don't build my own SDL version, i just include some private headers from SDL, because i need some Wayland features from SDL sources. But now, i try rewrite it without using private headers. 
OK, i will rebase it all to master branch.

**Comment by savegame** [comment]
looks like i add many uneeded binary data, maybe better make squash commit from that all? orr may be i need recreate branch with "clean" commits?

**Comment by CODeRUS** [comment]
git rebase -i is your friend ;)

**Comment by savegame** [comment]
need help: 
what i need to do? for disable MXAA feature in Sailfish? 
it seems in Sailfish, libGLESv2 dont have glRenderbufferStorageMultisample ( i try `nm -D libGLESv2.so|grep glRenderbufferStorageMultisample` )
now its compile well and work, but i have glitches, and error:
```
WARNING: _render_target_allocate: Cannot allocate mipmaps for 3D post processing effects
   At: drivers/gles2/rasterizer_storage_gles2.cpp:4916.
```

File: scene/main/scene_tree.cpp [hunk 1]
```diff
@@ -1232,6 +1278,9 @@ void SceneTree::_update_root_rect() {
 				WARN_PRINT("Font oversampling does not work in 'Viewport' stretch mode, only '2D'.");
 			}
 
+			// force landscape 
+			screen_size = Size2( screen_size.y, screen_size.x );
+			root->set_attach_to_screen_rect(Rect2(margin, screen_size));
```
Comment (savegame) [inline scene/main/scene_tree.cpp]: merge bug

**Comment by savegame** [comment]
hi everyone... 
i have troubles with master branch ) 
i merge and build Saiflish port from 3.1 branch to mater. On 3.1 my code works fine, but on master i have glitches:

| Sailfish Godot 3.1 | Sailfish Godot 3.2 |
| --- | --- |
| [  ](https://ibb.co/YpfHmdJ)  | [](https://ibb.co/4V4ZF9T) |
| [  ](https://ibb.co/ZN68YPf) | [  ](https://ibb.co/L6s71NH) |

as i know, reason is my - hack for SailfishOS landscape orientation, but seems now rotate only part of canvas on scene (from shader i mean), other items draw in default screen orientation  ... how can i understand in code, where i should draw finished scene buffer, and when i draw any canvas item (which should by rotated too, as all scene)
p.s. as i understand, in 3.1 all items was first drawn to scene buffer, then scene buffer was draw to screen, but in 3.2 items draw to screen directly? 
all my code with "hack" inside definitions **SAILFSIH_FORCE_LANDSCAPE**, and **gles2/shaders/canvas.glsl** 
@akien-mga, if you can, please tell where i should look, to fix my bug with drawing scene in landscape mode... 
3d now looks not works

**Comment by savegame** [comment]
now i see, i modified canvas.glsl shader, and every item on canvas draw with this shader... now i just need understand where in gles2_rasterizer draw final canvas buffer, and draw it with landscape UV modifications, and other canvas items draw normally.

**Comment by savegame** [comment]
:tada:  i found an answer! ) 
just need reset FORCE_LANDSCAPE in beggining of ""_canvas_item_render_commands"" function ?and all draw correctly now!

**Comment by savegame** [comment]
waiting for desigion  of maintainers :nerd_face:

**Comment by savegame** [comment]
@akien-mga,  is it need create new **PR** for this (i'm force rewrite my **3.1_sailfish_sdl2** branch with **master** branch), or just live it here as is?

**Comment by akien-mga** [comment]
It seems ok, the branch name doesn't matter much. What's important is the target branch, which is properly tracking `master`.

**Comment by savegame** [comment]
can any one restart travis check, dont know why mono is not build

**Comment by akien-mga** [comment]
Superseded by #38642.

**Pull request closed without merging**